hadoop-yarn-project/hadoop-yarn/hadoop-yarn-applications/hadoop-yarn-applications-distributedshell/src/test/java/org/apache/hadoop/yarn/applications/distributedshell/DistributedShellBaseTest.java

```diff
@@ -0,0 +1,557 @@
+    conf.set("yarn.log.dir", "target");
+    conf.setBoolean(YarnConfiguration.TIMELINE_SERVICE_ENABLED, true);
+    // mark if we need to launch the v1 timeline server
+    // disable aux-service based timeline aggregators
+    conf.set(YarnConfiguration.NM_AUX_SERVICES, "");
+    conf.setBoolean(YarnConfiguration.SYSTEM_METRICS_PUBLISHER_ENABLED, true);
+
+    conf.set(YarnConfiguration.NM_VMEM_PMEM_RATIO, "8");
+    conf.setBoolean(YarnConfiguration.NODE_LABELS_ENABLED, true);
+    conf.set("mapreduce.jobhistory.address",
+        "0.0.0.0:" + ServerSocketUtil.getPort(10021, 10));
+    // Enable ContainersMonitorImpl
+    conf.set(YarnConfiguration.NM_CONTAINER_MON_RESOURCE_CALCULATOR,
+        LinuxResourceCalculatorPlugin.class.getName());
+    conf.set(YarnConfiguration.NM_CONTAINER_MON_PROCESS_TREE,
+        ProcfsBasedProcessTree.class.getName());
+    conf.setBoolean(YarnConfiguration.NM_PMEM_CHECK_ENABLED, true);
+    conf.setBoolean(YarnConfiguration.NM_VMEM_CHECK_ENABLED, true);
+    conf.setBoolean(
+        YarnConfiguration.YARN_MINICLUSTER_CONTROL_RESOURCE_MONITORING,
+        true);
+    conf.setBoolean(YarnConfiguration.RM_SYSTEM_METRICS_PUBLISHER_ENABLED,
+        true);
+    conf.setBoolean(
+        YarnConfiguration.OPPORTUNISTIC_CONTAINER_ALLOCATION_ENABLED, true);
+    conf.setInt(YarnConfiguration.NM_OPPORTUNISTIC_CONTAINERS_MAX_QUEUE_LENGTH,
+        10);
+    conf.set(YarnConfiguration.RM_PLACEMENT_CONSTRAINTS_HANDLER,
+        YarnConfiguration.PROCESSOR_RM_PLACEMENT_CONSTRAINTS_HANDLER);
+    // ATS version specific settings
+    customizeConfiguration(conf);
+
+    if (yarnCluster == null) {
+      yarnCluster =
+          new MiniYARNCluster(getClass().getSimpleName(), 1, numNodeManagers,
+              1, 1);
+      yarnCluster.init(conf);
+      yarnCluster.start();
+    }
+    // wait for the node managers to register.
+    waitForNMsToRegister();
+
+    conf.set(
+        YarnConfiguration.TIMELINE_SERVICE_WEBAPP_ADDRESS,
+        MiniYARNCluster.getHostname() + ":"
+            + yarnCluster.getApplicationHistoryServer().getPort());
+
+    URL url = Thread.currentThread().getContextClassLoader().getResource(
+        "yarn-site.xml");
+    if (url == null) {
+      throw new RuntimeException(
+          "Could not find 'yarn-site.xml' dummy file in classpath");
+    }
+    Configuration yarnClusterConfig = yarnCluster.getConfig();
+    yarnClusterConfig.set(YarnConfiguration.YARN_APPLICATION_CLASSPATH,
+        new File(url.getPath()).getParent());
+    // write the document to a buffer (not directly to the file, as that
+    // can cause the file being written to get read -which will then fail.
+    ByteArrayOutputStream bytesOut = new ByteArrayOutputStream();
+    yarnClusterConfig.writeXml(bytesOut);
+    bytesOut.close();
+    // write the bytes to the file in the classpath
+    OutputStream os = new FileOutputStream(url.getPath());
+    os.write(bytesOut.toByteArray());
+    os.close();
+  }
```

Comment:
This overwrites target/test-classes/yarn-site.xml (copied from src/test/resources/yarn-site.xml on `mvn test` startup) by updated contents in order to propagate the configutaion to app containers. I guess the updated yarn-stie.xml is reloaded on running another test classes.

Resetting the yarn-site.xml by blank Configuration in DistributedShellBaseTest#tearDown (by the code like below) worked for me, while there might be better solution.
```
    URL url = Thread.currentThread().getContextClassLoader().getResource(
        "yarn-site.xml");
    OutputStream os = new FileOutputStream(url.getPath());
    new Configuration(false).writeXml(os);
    os.close();
```

Reply:
> This overwrites target/test-classes/yarn-site.xml (copied from src/test/resources/yarn-site.xml on `mvn test` startup) by updated contents in order to propagate the configutaion to app containers. I guess the updated yarn-stie.xml is reloaded on running another test classes.
> 
> Resetting the yarn-site.xml by blank Configuration in DistributedShellBaseTest#tearDown (by the code like below) worked for me, while there might be better solution.
> 
> ```
>     URL url = Thread.currentThread().getContextClassLoader().getResource(
>         "yarn-site.xml");
>     OutputStream os = new FileOutputStream(url.getPath());
>     new Configuration(false).writeXml(os);
>     os.close();
> ```

Thanks @iwasakims ! Actually the same code block exists in many unit tests (`serviceTestUtils`, `TestUnmanagedAMLauncher`, TestDynamometerInfra`, etc...).
should be the best practice to set the `YARN_SITE_CONFIGURATION_FILE` so that each Test class use a separate file? I assume that having one `yarn-site.xml` in classPath would cause problems running TestUnits in parallel.

```java
    yarnClusterConfig.writeXml(bytesOut);
    bytesOut.close();
    // write the bytes to the file in the classpath
    OutputStream os = new FileOutputStream(url.getPath());
    os.write(bytesOut.toByteArray());
```

Reply:
I pushed a fix  pushed fix keeping a backup in `BeforeClass` , then restore it in `AfterClass`. Probably we will need to do that fix in other classes too.